Add native HTTPS

mlegner opened the pull request:
This PR addresses issue #23 and adds the option to natively serve the website via HTTPS.
If enabled, the certificate is autogenerated via Let's Encrypt.
A registered domain is needed to enable the option.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion-coord/74)
<!-- Reviewable:end -->

ercanucan commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[config/config.go, line 46 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KvaYgutSO1yhFMMhWZH:-KvaYguuSf9NYazLJEGd:b6555pi) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/b71e9e01f6a8d56511d708261ec33754e5817232/config/config.go#L46)):*
> ```Go
> 	DB_NAME               = goconf.AppConf.String("db.name")
> 	DB_HOST               = goconf.AppConf.String("db.host")
> 	DB_PORT               = goconf.AppConf.DefaultInt("db.port", 3306)
> ```

Why are these changes necessary for this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74)*
<!-- Sent from Reviewable.io -->

mlegner commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[config/config.go, line 46 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KvaYgutSO1yhFMMhWZH:-KvfNOwt3NZPCtgauCie:bn2jt8h) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/b71e9e01f6a8d56511d708261ec33754e5817232/config/config.go#L46)):*
<details><summary><i>Previously, ercanucan (ercanucan) wrote…</i></summary><blockquote>

Why are these changes necessary for this PR?
</blockquote></details>

They are not at all necessary for this specific PR. I just noticed that we do not do any error handling in the config package and thought that it would make sense to use default values if none are given or they are incorrectly formatted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74)*
<!-- Sent from Reviewable.io -->

chaehni commented:
Made a pass.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[main.go, line 229 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqIZkR1NiHZ4-5T4z7:-KyqIZkR1NiHZ4-5T4z8:b7h1koc) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L229)):*
> ```Go
> 	if config.HTTP_ENABLE_HTTPS {
> 		// redirect HTTP traffic to HTTPS
> 		go http.ListenAndServe(":80", http.HandlerFunc(
> ```

Here, we are redirecting traffic from all host addresses to HTTPS. This blocks the use of port 80 for other addresses. Is this intended?
I'm not completely sure, but I think ListenAndServe() calls Serve() which in turn spawns a new goroutine by itself. 

---

*[main.go, line 234 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqEV7v18Mg2KvgfNWH:-KyqEV7w-0sXiFkJXZqJ:b-4ho7s0) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L234)):*
> ```Go
> 			}))
> 
> 		// listen to HTTP requests
> ```

'listen to HTTPS requests'

---

*[config/config.go, line 46 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KvaYgutSO1yhFMMhWZH:-KyqO8zf0vfcWpeuI7aX:b-etepu2) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/b71e9e01f6a8d56511d708261ec33754e5817232/config/config.go#L46)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

They are not at all necessary for this specific PR. I just noticed that we do not do any error handling in the config package and thought that it would make sense to use default values if none are given or they are incorrectly formatted.
</blockquote></details>

The default values were removed because having config values in two places  kept causing confusion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-:-KyqPiwIfqi7Ekmt6KIS:bn32ixk)*
<!-- Sent from Reviewable.io -->

mlegner commented:
Thank you for your comments. I addressed them and rebased my PR. If you are satisfied, we can merge it.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[main.go, line 229 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqIZkR1NiHZ4-5T4z7:-KytdG09KqwB9eqsDsB5:b-oabiys) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L229)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

Here, we are redirecting traffic from all host addresses to HTTPS. This blocks the use of port 80 for other addresses. Is this intended?
I'm not completely sure, but I think ListenAndServe() calls Serve() which in turn spawns a new goroutine by itself. 
</blockquote></details>

Yes, that is intended. The idea is that users can access the website via either HTTP or HTTPS but will be served via HTTPS in any case. As we are using a  dedicated server, this should pose no problem.
Indeed a goroutine is spawned in the function `Serve` in `net/http/server.go`. However, this doesn't seem to be enough, at least it didn't work in my tests, when I tried it without spawning a goroutine myself.

---

*[main.go, line 234 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqEV7v18Mg2KvgfNWH:-KytdkOGDwYz_aWBrUyW:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L234)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

'listen to HTTPS requests'
</blockquote></details>

Done.

---

*[config/config.go, line 46 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KvaYgutSO1yhFMMhWZH:-Kytdlie_uGa2DLPen-b:b-e1mvtm) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/b71e9e01f6a8d56511d708261ec33754e5817232/config/config.go#L46)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

The default values were removed because having config values in two places  kept causing confusion.
</blockquote></details>

OK, I didn't think about this rationale.
However, I still believe that it makes sense to have default values in the config module in order to avoid errors. Still, I have removed these default values from this PR and consider making a new PR where we can discuss this further.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-:-KyteQUlXq17a2obfQN_:bpff7qm)*
<!-- Sent from Reviewable.io -->

chaehni commented:
Made another pass. Just one comment left.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[main.go, line 229 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqIZkR1NiHZ4-5T4z7:-KyueYTljDnXQvDhtDPp:bl1so26) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L229)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Yes, that is intended. The idea is that users can access the website via either HTTP or HTTPS but will be served via HTTPS in any case. As we are using a  dedicated server, this should pose no problem.
Indeed a goroutine is spawned in the function `Serve` in `net/http/server.go`. However, this doesn't seem to be enough, at least it didn't work in my tests, when I tried it without spawning a goroutine myself.
</blockquote></details>

I think we can have the desired behavior of redirecting all traffic to HTTPS for SCION Coord and still leave port 80 untouched for other addresses by simply adding the host address in the same way we've done it for the non-HTTPS case below (line 247). Either way, we should be consistent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-:-KyuhFhR8LrND4KtgmFe:b9rhuhq)*
<!-- Sent from Reviewable.io -->

mlegner commented:
Thanks. Please see the comment below.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[main.go, line 229 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqIZkR1NiHZ4-5T4z7:-KyulNnYtg-vo0CZoVMZ:bf5zveh) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L229)):*
<details><summary><i>Previously, chaehni (chaehni) wrote…</i></summary><blockquote>

I think we can have the desired behavior of redirecting all traffic to HTTPS for SCION Coord and still leave port 80 untouched for other addresses by simply adding the host address in the same way we've done it for the non-HTTPS case below (line 247). Either way, we should be consistent.
</blockquote></details>

I think, this is unfortunately not possible as the automatic certificate management of letsencrypt needs the webserver to listen at ports 80 and 443. For a production setup, this is anyway what one would do while for development one can simply continue using the http setup.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-:-Kyulp5LDhlgniytr04O:b-wa06c0)*
<!-- Sent from Reviewable.io -->

chaehni commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[main.go, line 229 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-KyqIZkR1NiHZ4-5T4z7:-Kyux2zOUInv_sko24jB:b-lo8pcr) ([raw file](https://github.com/netsec-ethz/scion-coord/blob/5bee172b004daaff7cbe62a2fe2c691619453914/main.go#L229)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

I think, this is unfortunately not possible as the automatic certificate management of letsencrypt needs the webserver to listen at ports 80 and 443. For a production setup, this is anyway what one would do while for development one can simply continue using the http setup.
</blockquote></details>

I guess as long as we are not running other servers on the deployment machine it's not going to be a problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-:-Kyux1otDDck5jlI2Vr0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mlegner commented:
@ercanucan: Could you check and close your remaining issues then this can be merged.

ercanucan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74#-:-Kz4UrNwmST-xfygvNAE:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
Reviewed 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-coord/74)*
<!-- Sent from Reviewable.io -->
